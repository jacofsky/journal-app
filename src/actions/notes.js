import { addDoc, collection, deleteDoc, doc, updateDoc } from "firebase/firestore";
import Swal from "sweetalert2";
import { db } from "../firebase/firebase-config";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

// react-journal

export const startNewNote = () => {
    return async (dispatch, getState) => {
        
        const uid = getState().auth.uid;
        console.log(uid) 

        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        const doc = await addDoc(collection(db, `${uid}`, "journal/notes"), newNote)

        console.log("Document written with ID: ", doc);

        dispatch(activeNote(doc.id, newNote))
        dispatch(addNetNote(doc.id, newNote))


    
    }
}

export const activeNote = (id, note) => ({

    type: types.notesActive,
    payload: {
        id, 
        ...note
    }

}) 

export const addNetNote = (id, note) => ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note
    }
})

export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid)
        dispatch(setNotes(notes))
    }
}

export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: notes
})

export const startSaveNote = (note) => {
    return async(dispatch, getState) => {
        const uid = getState().auth.uid;

        if(!note.url){
            delete note.url
        }

        const noteToFirestore = {...note}
        delete noteToFirestore.id;

        await updateDoc(doc(db, `${uid}/journal/notes/${note.id}`), noteToFirestore)

        dispatch(refreshNote(note.id, noteToFirestore))
        Swal.fire({
            icon: 'success',
            title: 'Saved',
            text: note.title,
            confirmButtonColor: '#5cc5bc'
        })
    }
}

export const refreshNote = (id, note) => ({
    type: types.notesUpdated,
    payload: {
        id,
        note: {
            id,
            ...note
        }
    }
})

export const startUploading = (file) => {
    return async(dispatch, getState) => {

        const {active:activeNote} = getState().notes;

        
        Swal.fire({
            title: 'Uploading...',
            text: 'Please wait...',
            allowOutsideClick: false,
            didOpen: () => {
                Swal.showLoading()
            }
        })
        
        const imgUrl = await fileUpload(file)  
        activeNote.url = imgUrl;
        
        dispatch(startSaveNote(activeNote)) 
        
        
        Swal.close()  
        
        
    }
}

export const startDeleating = (id) => {
    return async(dispatch, getState) => {

        const uid = getState().auth.uid;

        await deleteDoc(doc(db, `${uid}/journal/notes/${id}`))

        dispatch(deleteNote(id))
    }
}

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
})

export const logoutCleaning = () => ({
    type: types.notesLogoutCleaning
})