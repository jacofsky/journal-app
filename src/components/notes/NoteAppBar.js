import React from 'react'

const NoteAppBar = () => {
    return (
        <div className='notes__appbar pt-3'>
            <span>1 enero 2022</span>

            <div className='notes__appbar--buttons'>
                <button className='btn btn-outline-light me-1'>
                    Picture
                </button>

                <button className='btn btn-outline-light ms-1'>
                    Save
                </button>
            </div>
        </div>
    )
}

export default NoteAppBar
