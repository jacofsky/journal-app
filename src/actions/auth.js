import Swal from 'sweetalert2'

import { createUserWithEmailAndPassword, getAuth, signInWithEmailAndPassword, signInWithPopup, updateProfile } from 'firebase/auth';
import { googleAuthProvider } from '../firebase/firebase-config';
import {types} from '../types/types'
import { finishLoading, startLoading } from './ui';
import { logoutCleaning } from './notes';

export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {

        dispatch(startLoading())
        const auth = getAuth()
        signInWithEmailAndPassword(auth, email, password)
            .then(({user}) => {      
                dispatch(login(user.uid, user.displayName))
                dispatch(finishLoading())
            })
            .catch(e =>{ 
                console.log(e) 
                dispatch(finishLoading())
                Swal.fire({
                    icon: 'error',
                    title: 'Oops...',
                    text: 'Check your email or password',
                    confirmButtonColor: '#5cc5bc'
                  })
            })
    
    }
}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {

        const auth = getAuth()
        createUserWithEmailAndPassword(auth, email, password)
            .then(async ({user}) => {
                await updateProfile(auth.currentUser, {displayName: name})
                
                dispatch(login(user.uid, user.displayName))
            })
            .catch (e => Swal.fire({
                icon: 'error',
                title: 'Oops...',
                text: 'The email has already an account',
                confirmButtonColor: '#5cc5bc'
              }))
    
    }
}

export const startGoogleLogin = () => {
    
    return (dispatch) => {
        
        const auth = getAuth()
        signInWithPopup(auth, googleAuthProvider)
            .then(({user}) => {
                dispatch(login(user.uid, user.displayName))
            })
            
    }

}

export const login = (uid, displayName) => ({
        
    type: types.login,
    payload: {
        uid,
        displayName
    }
    
})

export const startLogout = () => {
    return async(dispatch) => {
        const auth = getAuth()
        await auth.signOut()

        dispatch(logout())
        dispatch(logoutCleaning())
    }
}

export const logout = () => ({
    
    type: types.logout

})