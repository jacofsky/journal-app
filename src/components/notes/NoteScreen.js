import React from 'react'
import NoteAppBar from './NoteAppBar'

const NoteScreen = () => {
    return (
        <div className='notes__main-content'>
            <NoteAppBar />
            <div className="notes__content">
                    <input type="text" placeholder='Some awesome title' className='notes__title--input' autoComplete='off' />
                    <textarea placeholder='What happened today' className='notes__title--textarea' autoComplete='off' ></textarea>
                    
                    <div className="notes__image">
                        <img src="https://cutt.ly/qUAJKZd" alt="Los simpsons" />
                    </div>
            </div>
        </div>
    )
}

export default NoteScreen
