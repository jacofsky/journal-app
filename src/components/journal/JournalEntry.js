import React from 'react'
import moment from 'moment';
import { useDispatch } from 'react-redux';
import { activeNote } from '../../actions/notes';

const JournalEntry = ({id, date, title, body, url}) => {

    const dispatch = useDispatch()

    const noteDate = moment(date)

    const handleEntryClick = () => {
        dispatch(activeNote(id, {date, title, body, url}))
    }

    return (
        <div className='journal__entry mb-2 pointer animate__animated animate__fadeIn' onClick={handleEntryClick}>
            
            {
                url &&

                <div className="journal__entry--picture h-auto" style={{
                    backgroundSize: 'cover',
                    backgroundImage: `url(${url})`,
                    backgroundPosition: 'center'
                }}></div>
            }
            

            <div className="journal__entry--body">
                <p className="journal__entry--title">
                    {title}
                </p>
                <p className="journal__entry--content">
                    {body} 
                </p>
                
            </div>

            <div className="journal__entry-date-box ms-auto pe-3">
                <span>{noteDate.format('dddd')}</span> <h4 className='m-0'>{noteDate.format('Do')}</h4>
            </div>
        </div>
    )
}

export default JournalEntry
