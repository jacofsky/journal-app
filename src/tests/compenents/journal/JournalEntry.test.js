import React from "react"
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import JournalEntry from "../../../components/journal/JournalEntry"

import { mount } from "enzyme"
import { Provider } from "react-redux"
import { activeNote } from "../../../actions/notes"



jest.mock('../../../actions/notes', () => ({
    activeNote: jest.fn()

}))



const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initState = {}

let store = mockStore(initState)
store.dispatch = jest.fn()

const nota = {
    id: 10,
    date: 1,
    title: 'Hola',
    body: 'Mundo',
    url: 'https://algunlugar.com/foto.jpg'
}

const wrapper = mount(
    <Provider store={store}>
        <JournalEntry {...nota}/>
    </Provider>
)

describe('Test in <JournalEntry />', () => {

    test('should show snapshot', () => { 

        expect(wrapper).toMatchSnapshot()

    })

    test('should active note', () => { 

        wrapper.find('.journal__entry').simulate('click')

        expect(store.dispatch).toHaveBeenCalledWith(
            activeNote(nota.id, {...nota})
        )

    })

})