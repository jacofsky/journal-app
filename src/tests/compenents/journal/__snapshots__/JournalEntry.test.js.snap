// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <JournalEntry /> should show snapshot 1`] = `
<div
  className="journal__entry mb-2 pointer animate__animated animate__fadeIn"
  onClick={[Function]}
>
  <div
    className="journal__entry--picture h-auto"
    style={
      Object {
        "backgroundImage": "url(https://algunlugar.com/foto.jpg)",
        "backgroundPosition": "center",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry--body"
  >
    <p
      className="journal__entry--title"
    >
      Hola
    </p>
    <p
      className="journal__entry--content"
    >
      Mundo
    </p>
  </div>
  <div
    className="journal__entry-date-box ms-auto pe-3"
  >
    <span>
      Wednesday
    </span>
     
    <h4
      className="m-0"
    >
      31st
    </h4>
  </div>
</div>
`;
